subprojects/dependency-management/src/main/java/org/gradle/api/internal/artifacts/ivyservice/ivyresolve/verification/WriteDependencyVerificationFile.java

```diff
@@ -0,0 +1,169 @@
+        conf.setLenient(true);
+    };
+
+    private final DependencyVerifierBuilder verificationsBuilder = new DependencyVerifierBuilder();
+    private final File buildDirectory;
+    private final BuildOperationExecutor buildOperationExecutor;
+    private final Map<CachedChecksum, String> cachedChecksums = Maps.newConcurrentMap();
+
+    public WriteDependencyVerificationFile(File buildDirectory, BuildOperationExecutor buildOperationExecutor) {
+        this.buildDirectory = buildDirectory;
+        this.buildOperationExecutor = buildOperationExecutor;
+    }
+
+    @Override
+    public ModuleComponentRepository overrideDependencyVerification(ModuleComponentRepository original) {
+        return new DependencyVerifyingModuleComponentRepository(original, this);
+    }
+
+    @Override
+    public void buildFinished(Gradle gradle) {
+        File verifFile = DependencyVerificationOverride.dependencyVerificationsFile(buildDirectory);
+        try {
+            if (verifFile.exists()) {
+                LOGGER.info("Found dependency verification metadata file, updating");
+                DependencyVerificationsXmlReader.readFromXml(new FileInputStream(verifFile), verificationsBuilder);
+            }
+            gradle.allprojects(WriteDependencyVerificationFile::resolveAllConfigurationsAndForceDownload);
+            DependencyVerificationsXmlWriter.serialize(
+                verificationsBuilder.build(),
+                new FileOutputStream(verifFile)
+            );
+        } catch (IOException e) {
+            throw UncheckedException.throwAsUncheckedException(e);
+        }
+    }
+
+    @Override
+    public void onArtifact(ModuleComponentArtifactIdentifier id, File file) {
+        addChecksum(id, file, ChecksumKind.sha1);
+        addChecksum(id, file, ChecksumKind.sha512);
+    }
+
+    private void addChecksum(ModuleComponentArtifactIdentifier id, File file, ChecksumKind kind) {
+        buildOperationExecutor.run(new RunnableBuildOperation() {
+            @Override
+            public void run(BuildOperationContext context) {
+                verificationsBuilder.addChecksum(id, kind, () -> createHash(file, kind));
+            }
+
+            @Override
+            public BuildOperationDescriptor.Builder description() {
+                return BuildOperationDescriptor.displayName("Generate " + kind + " checksum for " + id);
+            }
+        });
+
+    }
+
+    private String createHash(File file, ChecksumKind kind) {
+        return cachedChecksums.computeIfAbsent(new CachedChecksum(file, kind), key -> HashUtil.createHash(file, kind.getAlgorithm()).asHexString());
+    }
+
+    private static void resolveAllConfigurationsAndForceDownload(Project p) {
+        p.getConfigurations().all(cnf -> {
+            if (((DeprecatableConfiguration) cnf).canSafelyBeResolved()) {
+                try {
+                    resolveAndDownloadExternalFiles(cnf);
```

Comment:
I wonder about the safety of this, since we are not supposed to resolve configurations outside of a project lock. But I assume that since `buildFinished` is single threaded and issue by Gradle, we are fine.
At the same time, in case the build itself did not resolve anything, this step could be extremely costly for large builds.
So should we attempt it in parallel?
But of course doing that re-opens the safety issue.

Reply:
Is this really unsafe? It happens within a `allprojects` block so I would assume the project lock is acquired. I agree it would be nice to do in parallel, I'm already experimenting with this. 

Serially on the Gradle build itself, it takes [around ~12s to build the checksums](https://e.grdev.net/s/f4auio26qbvj6/dependencies) for 2726 configurations and 500 dependencies.